internal/resource/http/http_test.go

```diff
@@ -311,6 +312,88 @@ func createResource(t *testing.T, artistHandler resourceHandler) {
 	require.Equal(t, http.StatusOK, rsp.Result().StatusCode)
 }
 
+func TestResourceReadHandler(t *testing.T) {
+	aclResolver := &resourceSvc.MockACLResolver{}
+	aclResolver.On("ResolveTokenAndDefaultMeta", testACLTokenArtistReadPolicy, mock.Anything, mock.Anything).
+		Return(svctest.AuthorizerFrom(t, demo.ArtistV1ReadPolicy, demo.ArtistV2ReadPolicy), nil)
+	aclResolver.On("ResolveTokenAndDefaultMeta", testACLTokenArtistWritePolicy, mock.Anything, mock.Anything).
+		Return(svctest.AuthorizerFrom(t, demo.ArtistV1WritePolicy, demo.ArtistV2WritePolicy), nil)
+	aclResolver.On("ResolveTokenAndDefaultMeta", fakeToken, mock.Anything, mock.Anything).
+		Return(svctest.AuthorizerFrom(t, ""), nil)
+
+	client := svctest.RunResourceServiceWithACL(t, aclResolver, demo.RegisterTypes)
+
+	v2ArtistHandler := resourceHandler{
+		resource.Registration{
+			Type:  demo.TypeV2Artist,
+			Proto: &pbdemov2.Artist{},
+		},
+		client,
+		parseToken,
+		hclog.NewNullLogger(),
+	}
+
+	createResource(t, v2ArtistHandler)
+
+	t.Run("Read resource", func(t *testing.T) {
+		rsp := httptest.NewRecorder()
+		req := httptest.NewRequest("GET", "/demo/v2/artist/keith-urban?partition=default&peer_name=local&namespace=default&consistent", nil)
+
+		req.Header.Add("x-consul-token", testACLTokenArtistReadPolicy)
+
+		v2ArtistHandler.ServeHTTP(rsp, req)
+
+		require.Equal(t, http.StatusOK, rsp.Result().StatusCode)
+
+		var result map[string]any
+		require.NoError(t, json.NewDecoder(rsp.Body).Decode(&result))
+		// generation and uid are random
+		delete(result, "generation")
+		delete(result["id"].(map[string]interface{}), "uid")
+		expected := map[string]interface{}{
+			"data": map[string]interface{}{"genre": "GENRE_COUNTRY", "name": "Keith Urban"},
+			"id": map[string]interface{}{
+				"name": "keith-urban",
+				"tenancy": map[string]interface{}{
+					"namespace": "default",
+					"partition": "default",
+					"peerName":  "local",
+				},
+				"type": map[string]interface{}{
+					"group":        "demo",
+					"groupVersion": "v2",
+					"kind":         "Artist",
+				},
+			},
+			"metadata": map[string]interface{}{"foo": "bar"},
+			"version":  "1",
+		}
```

Comment:
What we really want to verify is if we are getting the same data that we persisted. To guarantee that we should compare the response of `createResource` directly with the read response.
See [here](https://github.com/hashicorp/consul/pull/18444/files#diff-ebd21d0842178b65ad1c6f73a1cfd51ffdff46a31bf5067cb1798611548155f6R280).

Reply:
makes sense, changed